### PR TITLE
Add versioning system to fast tokenizer files

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

Some changes cannot be done to the fast tokenizers file without breaking backward compatibility. This PR introduces a versioning system by allowing a model repo to contain multiple tokenizer files: the `tokenizer.json` is the default one and if one (or several) `tokenizer.x.y.z.json` exist, those files are used for the version x.y.z (of Transformers) and above.

cc @n1t0 as it should be helpful to solve that longstanding bug.

### CODE REVIEW

**Comment by julien-c** [comment]
might be cleaner if this worked in the other direction, i.e.

> multiple tokenizer files: the `tokenizer.json` is the default one, used in the most recent version of Transformers. If one or more `tokenizer-x.y.z.json` exist, those files are used for the version x.y.z (of Transformers) and below.

Makes more sense on the Hub side as well. What do you think?

**Comment by LysandreJik** [comment]
@julien-c this would break repositories that rely on `transformers` versions that are earlier than the first one that will have this feature.

Were we to update the `tokenizer.json` file to the new, "fixed" one, and add a new `tokenizer-x.x.x.json` file to be used by earlier versions of `transformers`, then we would have no way of telling all versions < `4.10.0` to use that version rather than the standard `tokenizer.json` file.

**Comment by julien-c** [comment]
I think your assertion depends on what kind of changes are made to the JSON files. If it's only new attributes for example I wouldn't expect older versions to break, but from what I understand you're actually talking about modifying the actual attributes?

**Comment by LysandreJik** [comment]
Yes, the attributes actually need to be modified. For example, see this issue: https://github.com/huggingface/transformers/issues/9633

There was an offset mappings bug, which needed to be patched. However, the issue lived in the `tokenizer.json` file itself - so the recommended way to patch this was for users to recompile that file, by passing the "slow" tokenizer files, and using the newer `tokenizers` version to generate the updated file.

I believe there are other issues, and there will be other issues as the libraries continue to evolve. Implementing this here allows us to ensure that the previous versions remain completely unaffected - while offering a way to patch models for future use.

File: src/transformers/file_utils.py [hunk 1]
```diff
@@ -1641,6 +1641,53 @@ def _resumable_file_manager() -> "io.BufferedWriter":
...
+    revision: Optional[str] = None,
+    use_auth_token: Optional[Union[bool, str]] = None,
+) -> List[str]:
+    """
+    Gets the list of files inside :obj:`path_or_repo`.
+
+    Args:
+        path_or_repo (:obj:`str` or :obj:`os.PathLike`):
+            Can be either the id of a repo on huggingface.co or a path to a `directory`.
+        revision(:obj:`str`, `optional`, defaults to :obj:`"main"`):
```
Comment (LysandreJik) [inline src/transformers/file_utils.py:1655]:
```suggestion
        revision (:obj:`str`, `optional`, defaults to :obj:`"main"`):
```

File: tests/test_tokenization_fast.py [hunk 2]
```diff
@@ -93,6 +95,51 @@ def test_training_new_tokenizer_with_special_tokens_change(self):
...
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            tokenizer.save_pretrained(tmp_dir)
+            json.dump(json_tokenizer, open(os.path.join(tmp_dir, "tokenizer.4.0.0.json"), "w"))
+
+            # This should pick the new tokenizer file as the version of Transformers is > 4.0.0
+            new_tokenizer = AutoTokenizer.from_pretrained(tmp_dir)
+            self.assertEqual(len(new_tokenizer), len(tokenizer) + 1)
+            json_tokenizer = json.loads(new_tokenizer._tokenizer.to_str())
+            self.assertIn("huggingface", json_tokenizer["model"]["vocab"])
```
Comment (LysandreJik) [inline tests/test_tokenization_fast.py:113]: Nice test!

File: tests/test_tokenization_fast.py [hunk 3]
```diff
@@ -93,6 +95,51 @@ def test_training_new_tokenizer_with_special_tokens_change(self):
...
+
+            # Will need to be adjusted if we reach v42 and this test is still here.
+            # Should pick the old tokenizer file as the version of Transformers is < 4.0.0
```
Comment (LysandreJik) [inline tests/test_tokenization_fast.py:116]: Let's think about it next century :)

**Review by LysandreJik — APPROVED** [review]
Ok cool! Played with it and it seems to be working well. LGTM, thanks a lot for working on it, this should really help with tokenizer files going forward!

**Comment by julien-c** [comment]
> Yes, the attributes actually need to be modified. For example, see this issue: #9633
> 
> There was an offset mappings bug, which needed to be patched. However, the issue lived in the `tokenizer.json` file itself - so the recommended way to patch this was for users to recompile that file, by passing the "slow" tokenizer files, and using the newer `tokenizers` version to generate the updated file.
> 
> I believe there are other issues, and there will be other issues as the libraries continue to evolve. Implementing this here allows us to ensure that the previous versions remain completely unaffected - while offering a way to patch models for future use.

going on a whim here, but what about using git branches to do this?

**Comment by sgugger** [comment]
The problem with a new branch is that we then can't have a new version of the model in a new git branch that has to be used with one tokenizer file if versions of Transformers are old, and another one if they are more recent. And it wouldn't be compatible with the sure selecting their own branch as well (though in that case they should make sure to have the right version with tokenizers file).

The key here (for more context) is that we have tokenizers that have a "wrong" tokenizer file for more recent versions of Tokenizers (controlled by the version of Transformers) because there was a bug in the conversion from slow to fast tokenizer script. We can't touch the main branch and the tokenizer.json file otherwise every code in production using those models will suddenly break (the changes are significant sadly).

File: tests/test_tokenization_fast.py (hunk 3 above)
Comment (julien-c) [reply tests/test_tokenization_fast.py:116]: when we near AGI this library will publish itself and who knows which version number it will pick

**Pull request merged**